flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/status/FlinkDeploymentStatus.java

```diff
@@ -55,4 +55,7 @@ public class FlinkDeploymentStatus extends CommonStatus<FlinkDeploymentSpec> {
 
     /** Information about the TaskManagers for the scale subresource. */
     private TaskManagerInfo taskManager;
+
+    /** Last observed generation of the FlinkDeployment. */
+    private Long observedGeneration;
```

Comment:
We do have this information available already, via the `lastReconciledSpec` field. Not sure whether we need to expose this as a top-level field.

CC @gyfora

Reply:
Exactly as @mxm said. 

Also the field inside the last reconciled spec has a slightly different and more correct semantics. I think what you did here (update the observed generation) doesn't make too much practical sense as the operator may fail before even seeing the spec. 

I see the value of having a top level `observedGeneration` field but it should replace the current field hidden inside the lastReconciledSpec with the same semantics.

Reply:
>We do have this information available already, via the lastReconciledSpec field. Not sure whether we need to expose this as a top-level field.

The idea is that the Kubernetes spec expects that this status field exists in this particular place. This type of PR allows for 3rd party tools to work well with the Flink Operator. (ie. our build tools internally now can communicate with the operator which is 🙏)

However, agreed with Gyula on how it should replace semantically what exists today.

Reply:
Thanks for the guidance - I've added `observedGeneration` to `CommonStatus` class and set it alongside the `lastReconciledSpec`. I've left the `generation` within the `lastReconciledSpec` intact in case there are existing client dependencies on it. Let me know about any additional changes to be made

Reply:
@justin-chen I would prefer to replace the implementation and call it out in the release notes instead. That way we end up with a cleaner codebase.

Any concerns from your end @mxm ?

Reply:
I would prefer to keep the field in lastReconciledSpec to avoid downstream breakage. At least for the next release.

Reply:
In that case we should set the new field next to where the old one is set so that the old logic is easy to remove later .

Reply:
And we should add a note of deprecation and a jira to track it. Although we don’t have any explicit guarantees here for the status especially for contents of json fields

Reply:
> we should set the new field next to where the old one is set so that the old logic is easy to remove later

@gyfora Not sure I understand, is the intention to duplicate the `observedGeneration` such that it's available via both `status.observedGeneration` (we need it at this top level) and `status.reconciliationStatus.lastReconciledSpec.resource_metadata`?

Given the status currently looks like this (with my current PR):
```
status:
    ...
    observedGeneration: 123
    reconciliationStatus:
    lastReconciledSpec: '{
        "spec": {
        ...
        },
        "resource_metadata": {
            "apiVersion": "flink.apache.org/v1beta1",
            "metadata": {   // io.fabric8.kubernetes.api.model.ObjectMeta
                "generation": 123
            },
            "firstDeployment": false
        }
    }'
    lastStableSpec: '{...}'
    reconciliationTimestamp:
    state: DEPLOYED
```

Reply:
I think you misunderstand. Let's leave the `generation` field in the lastReconciledSpec where it is now for backward compatibility. 

However instead of getting it from there and setting it in the top level status, set it in the top level status in the same step where we set in the lastReconciledSpec. That way the next person doesn't have to rewrite the entire logic to remove the deprecated meta field from lastReconciledSpec.

Reply:
In essence, the current code already reflected the desired functionality, we just need to move the code to set the top-level generation id to where we initially set it. I hope that's easily doable. If not, let us know.

Reply:
👍 I have updated the PR to set `status.observedGeneration` using the [same source of truth](https://github.com/apache/flink-kubernetes-operator/blob/main/flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/reconciler/ReconciliationMetadata.java#L44) as `status.reconciliationStatus.lastReconciledSpec.resource_metadata.generation`, such that we can later remove the latter field without the `observedGeneration` depending on it.

The new field is set in `updateStatusForSpecReconciliation`  method alongside the `lastReconciledSpec`.